### PR TITLE
Koa 4892 android add large button variant

### PR DESCRIPTION
**Pull request by bvitaliyg**
+ [ ] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-android/blob/main/CONTRIBUTING.md)

Remember to include the following changes:
+ [ ] `UNRELEASED.md`
+ [ ] Tests
+ [ ] Docs (either update [backpack-docs](https://github.com/Skyscanner/backpack-docs) now, or create a follow up ticket)

_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/main/CODE_REVIEW_GUIDELINES.md)_

### CODE REVIEW

File: Backpack/src/main/java/net/skyscanner/backpack/button/BpkButton.kt [hunk 1]
```diff
@@ -32,23 +32,26 @@ import net.skyscanner.backpack.button.internal.ButtonStyles
...
 open class BpkButton(
   context: Context,
   attrs: AttributeSet?,
   defStyleAttr: Int,
   type: Type,
+  private val format: BpkButtonFormat = BpkButtonFormat.Small,
 ) : BpkButtonBase(context, attrs, defStyleAttr) {
 
-  constructor(context: Context) : this(context, null, 0, Type.Primary)
+  constructor(context: Context) : this(context, null, 0, Type.Primary, BpkButtonFormat.Small)
```
Comment (bvitaliyg) [inline Backpack/src/main/java/net/skyscanner/backpack/button/BpkButton.kt:48]: BpkButtonSize causes compilation error, as some of the libraries has app:buttonSize property already.

File: app/src/main/java/net/skyscanner/backpack/demo/stories/LoadingButtonStory.kt [hunk 2]
```diff
@@ -21,25 +21,30 @@ package net.skyscanner.backpack.demo.stories
...
-          child.postDelayed(
-            {
-              child.loading = false
-            },
-            2500
-          )
+      when (child) {
+        is ViewGroup -> makeButtonsLoadeable(child)
+        is BpkButton -> child.setOnClickListener {
+          viewLifecycleOwner.lifecycleScope.launch {
```
Comment (bvitaliyg) [inline app/src/main/java/net/skyscanner/backpack/demo/stories/LoadingButtonStory.kt:43]: This is why I added androidXLifecycleKtx

File: Backpack/src/main/java/net/skyscanner/backpack/button/BpkButtonLink.kt [hunk 3]
```diff
@@ -93,5 +94,6 @@ open class BpkButtonLink @JvmOverloads constructor(
       (resources.getDimensionPixelSize(R.dimen.bpkBorderSizeLg) / 2)
 
     setPadding(0, paddingVertical, 0, paddingVertical)
+    BpkText.getFont(this.context, BpkText.TextStyle.Label2).applyTo(this)
```
Comment (marianeum) [inline Backpack/src/main/java/net/skyscanner/backpack/button/BpkButtonLink.kt:97]: link also needs the large variant, though

File: Backpack/src/main/java/net/skyscanner/backpack/button/BpkButton.kt [hunk 4]
```diff
@@ -187,47 +211,23 @@ open class BpkButton(
...
+  enum class Type {
+    Primary,
+    Secondary,
+    Featured,
+    Destructive,
+    PrimaryOnDark,
+    PrimaryOnLight,
+    ;
+
+    internal companion object
```
Comment (bvitaliyg) [inline Backpack/src/main/java/net/skyscanner/backpack/button/BpkButton.kt:223]: This is used as a point for static extensions

File: Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt [hunk 5]
```diff
@@ -19,30 +19,61 @@
...
-// mainly exists here for compatibility reasons
+internal const val ICON_POSITION_START = 0
+internal const val ICON_POSITION_END = 1
+internal const val ICON_POSITION_ICON_ONLY = 2
+
 abstract class BpkButtonBase internal constructor(
   context: Context,
   attrs: AttributeSet?,
-  defStyleAttr: Int
-) : BpkButtonWithIcon(context, attrs, defStyleAttr) {
```
Comment (bvitaliyg) [inline Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt:33]: BpkButtonWith icon had alsmost no userful code after migrating to MaterialButton, so I merged it.

File: Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt [hunk 6]
```diff
@@ -19,30 +19,61 @@
...
+  }
+
   init {
     maxLines = 1
     gravity = Gravity.CENTER
     ellipsize = TextUtils.TruncateAt.END
     isClickable = isEnabled
     iconSize = resources.getDimensionPixelSize(R.dimen.bpkSpacingBase)
     backgroundTintList = null
-    font.applyTo(this)
```
Comment (bvitaliyg) [inline Backpack/src/main/java/net/skyscanner/backpack/button/internal/BpkButtonBase.kt:46]: BpkButtonLink applies the font manually instead

File: app/src/androidTest/java/net/skyscanner/backpack/button/BpkButtonBaseTest.kt [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import androidx.appcompat.content.res.AppCompatResources
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import net.skyscanner.backpack.BpkSnapshotTest
+import net.skyscanner.backpack.demo.R
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+
+@RunWith(AndroidJUnit4::class)
+class BpkButtonBaseTest : BpkSnapshotTest() {
```
Comment (bvitaliyg) [inline app/src/androidTest/java/net/skyscanner/backpack/button/BpkButtonBaseTest.kt:31]: This is the tests speciefic to view button implementation which check for speciefic scenarios.

File: Backpack/src/main/java/net/skyscanner/backpack/button/BpkButtonLink.kt (hunk 3 above)
Comment (bvitaliyg) [reply Backpack/src/main/java/net/skyscanner/backpack/button/BpkButtonLink.kt:97]: Lets do in in another ticket

Comment (marianeum) [reply Backpack/src/main/java/net/skyscanner/backpack/button/BpkButtonLink.kt:97]: I'd rather tackle it as part of the same change as it'd probably have breaking implications to do so later - i.e. moving the enum to the base class

File: docs/view/Button/README.md [hunk 8]
```diff
@@ -34,6 +34,35 @@ BpkButton(context,BpkButton.Type.Primary).apply {
 }
 ```
 
+### Large
```
Comment (marianeum) [inline docs/view/Button/README.md:37]: can we update the screenshots/demo app as well to include a large variant?

Comment (bvitaliyg) [reply docs/view/Button/README.md:37]: It's done already :)

Comment (marianeum) [reply docs/view/Button/README.md:37]: ah sorry this was meant to be on the link readme - not buttons. buttons exist ,but not link

**Review by marianeum — APPROVED** [review]

**Pull request merged**